File: 0_数组/原地操作/retain.ts
/**
 * 保留数组中满足条件的元素.
 * 原地操作.
 */
function retain(arr: any[], f: (index: number) => boolean): void {
  let ptr = 0
  for (let i = 0; i < arr.length; i++) {
    if (f(i)) arr[ptr++] = arr[i]
  }
  arr.length = ptr
}

export { retain }

File: 0_数组/原地操作/retain.go
package main

import "fmt"

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	Retain(&arr, func(index int) bool { return arr[index]%2 == 0 })
	fmt.Println(arr)
}

func Retain[T any](arr *[]T, f func(index int) bool) {
	ptr := 0
	n := len(*arr)
	for i := 0; i < n; i++ {
		if f(i) {
			(*arr)[ptr] = (*arr)[i]
			ptr++
		}
	}
	*arr = (*arr)[:ptr:ptr]
}
